src/Stratis.Features.FederatedPeg/CheckCollateralFullValidationRule.cs

```diff
@@ -0,0 +1,45 @@
+﻿using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Stratis.Bitcoin.Consensus.Rules;
+using Stratis.Bitcoin.Features.PoA;
+using Stratis.Bitcoin.Interfaces;
+
+namespace Stratis.Features.FederatedPeg
+{
+    /// <summary>Ensures that collateral requirement on counterpart chain is fulfilled for the federation member that produced a block.</summary>
+    /// <remarks>Ignored in IBD.</remarks>
+    public class CheckCollateralFullValidationRule : FullValidationConsensusRule
+    {
+        private readonly IInitialBlockDownloadState ibdState;
+
+        private readonly ICollateralChecker collateralChecker;
+
+        private readonly ISlotsManager slotsManager;
+
+        public CheckCollateralFullValidationRule(IInitialBlockDownloadState ibdState, ICollateralChecker collateralChecker, ISlotsManager slotsManager)
+        {
+            this.ibdState = ibdState;
+            this.collateralChecker = collateralChecker;
+            this.slotsManager = slotsManager;
+        }
+
+        public override Task RunAsync(RuleContext context)
+        {
+            if (this.ibdState.IsInitialBlockDownload())
```

Comment:
Because of the calculation for IBD always ending at N hours / blocks away from the present, couldn't this cause issues.

ie following scenario:
- Masternode owner has collateral on main chain, mines block on PoA sidechain. Waits a small period, a few blocks. Then moves collateral to new address, nodes start thinking he doesn't have collateral requirement.
- Now anyone syncing on sidechain is going to get out of IBD, do the collateral check for the last ~100 blocks or so, and see that this masternode owner doesn't have the collateral. So block is invalid. Can't sync fully.

Am I missing anything here?

Reply:
Yes, that is correct. IBD nodes are vulnerable during this period and will have to resync in case federation member that is being kicked decides to be malicious. In order to do so this fed member will have to track the moment when majority voted to kick him, since that moment he has maxReorg blocks till this happens to mine a block on sidechain and move collateral to another address and to perform this attack.

Reply:
I think it's possible to fix but it's not a big deal because there is no monetary incentive to perform this attack, also you can't do it anonymously and you need to modify client to perform it. And most importantly: attack affects only gateway nodes, not normal users, so what is the chance that block producing node will do IBD at the time of the attack? We expect all of them being online all the time so why would they be in IBD? And even if attack occurs- ok, it's not a big deal to resync.

Reply:
I wonder if this should be a full validation rule, probably the better place to put it is Integrity (or partial) What if the mainchain node falls behind and the collateral check is not accurate from this nodes perspective but the producer of the block and the rest of the network see it as correct, the node will ban the block and for away.

Reply:
>I wonder if this should be a full validation rule, probably the better place to put it is Integrity (or partial).

If it's not a FV then the check can happen at any time, not just before we add the block to the consensus chain. 

>What if the mainchain node falls behind and the collateral check is not accurate from this nodes perspective

Then it's a problem. Node should never fall behind since we are relying on it.